Focal Point Picker

jeffersonrabb opened the pull request:
## Description
This PR creates a Focal Point Picker component and implements it in the `Cover` block. Here is the rationale for this functionality: when dealing with large background images like the ones used in `Cover`, it's common to see undesirable crops, especially on smaller viewports. If the primary subject of an image happens to be dead center then it will display perfectly everywhere, but if it is off-center by even a little, key visual information can be cropped out.

The Focal Point Picker component allows the user to visually select the ideal center point, which is then returned as a pair of percentage-like coordinates (e.g. `{ x: 0.1, y: 0.9 }`) which can easily be converted into `background-position` attributes and applied to the element with the background image. 

### Video and Parallax

The Focal Point Picker only works if an image is being used, and if `Fixed Background` is disabled. For now, the same approach won't be applied to videos, and Parallaxed images don't really require a focal point since their crop changes as the user scrolls.

### Legacy Content

As part of this branch, `Cover` block saves the dimensions of images when selected, which is needed by the component. For instances of `Cover` block that pre-date this branch there will be no dimensions data, and therefore the focal point picker is suppressed. Because of this approach there was no need to add a Gutenberg deprecation definition.

## How has this been tested?

- Spin up local Gutenberg instance following instructions here: https://github.com/WordPress/gutenberg/blob/master/CONTRIBUTING.md
- Switch branch to `try/focal-point-picker`
- Create new post
- Add `Cover` block to post.
- Upload or select a large image, ideally one where the central focus is not off center - the further the better.
- Click on block, then show block settings. In the sidebar, click and drag the picker icon to the 
focal point of the image. Depending on the dimensions, you will probably see the image move around a little in the editor. 
- Save the post and preview. 

## Screenshots <!-- if applicable -->

Here are two screen captures demonstrating the use of the focal point picker:

https://cloudup.com/cIIvWAafA20

https://cloudup.com/cnL3p-eZvSq

## Types of changes

The Focal Point Picker is a new feature, which shouldn't affect anything, however the implementation in `Cover` could possibly break the block and should be tested thoroughly. Another important testing flow is to create a `Cover` block while on `master`, then branch switch to `try/focal-point-picker`. The block should remain exact as it was, with no focal point picker visible.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

jeffersonrabb committed:
Focal Point Picker, initial commit. Cover block implementation of Focal Point Picker for non-parallaxed image backgrounds.

jeffersonrabb committed:
Focal Point Picker README, initial commit.

jeffersonrabb committed:
Fixes for Cover blocks that pre-date the Focal Point Picker: 1) suppress focal point picker UI if image dimensions aren't available, 2) suppress background-position styling unless focal point is valid.

jeffersonrabb committed:
Small mistake in sample code.

jeffersonrabb committed:
README edits.

mtias commented:
Thanks @jeffersonrabb for the proposal and implementation! I think this will be a neat addition. Having it as a reusable component would also help others build blocks with equal capabilities (and a consistent experience).

jeffersonrabb committed:
Missing file.

MichaelArestad commented:
This is great!

designsimply commented:
@jasmussen, @jeffersonrabb is going to rebase and I was wondering if you might be able to help out with a design feedback review?

jeffersonrabb committed:
Include Focal Point Picker stylesheet in packages/components/src/style.scss instead of the components index.s.

jeffersonrabb committed:
Some conflict resolution in cover block's index.js.

jeffersonrabb commented:
Rebased.

gziolo commented:
I didn’t do very in-depth review of this PR, just initial pass. At first glance, it looks good. Thanks for working on it 👍

jeffersonrabb commented on packages/components/src/focal-point-picker/README.md:
```diff
@@ -0,0 +1,68 @@
+} );
+
+/* Example function to render the CSS styles based on Focal Point Picker value */
+const renderImageContainerWithFocalPoint = ( url, focalPoint ) => {
```
This is meant to illustrate how the percentage values from the component can be translated into a "real-world" CSS `background-position` definition. The relevant line in Cover block is:

https://github.com/WordPress/gutenberg/blob/ded1017374c1bb815ba02ea7c04637be5dbf981b/packages/block-library/src/cover/index.js#L241

It would probably make sense for me to use the actual example instead of the hypothetical, but I'd prefer to hold off on updating the README until the PR is ready to merge, in case there are changes still to come. Would this be a reasonable approach?

jeffersonrabb committed:
Regrouping of component dependencies.

jeffersonrabb commented on packages/components/src/focal-point-picker/index.js:
```diff
@@ -0,0 +1,159 @@
+import { withInstanceId, compose } from '@wordpress/compose';
+import BaseControl from '../base-control/';
+import withFocusOutside from '../higher-order/with-focus-outside';
+import classnames from 'classnames';
```
Resolved in https://github.com/WordPress/gutenberg/pull/10925/commits/97fcd04d05ce2ae039f891225a6a1fb781edfab2

jeffersonrabb commented on packages/components/src/focal-point-picker/README.md:
```diff
@@ -0,0 +1,68 @@
+		y: 0.5
+	},
+} )( ( { focalPoint, setState } ) => { 
+	const url = '/path/to/image';
```
Are there examples of valid url parameters in any other Gutenberg components? I just did a quick scan looking for one and came up blank, but maybe you can point me to a relevant one?

gziolo replied:
See how `ResponsiveWrapper` is documented:
https://github.com/WordPress/gutenberg/blob/master/packages/components/src/responsive-wrapper/README.md

This is how it renders in Calypso DevDocs:
https://wpcalypso.wordpress.com/devdocs/gutenberg-components/responsive-wrapper

youknowriad commented on packages/block-library/src/cover/index.js:
```diff
@@ -67,6 +68,12 @@ const blockAttributes = {
+	},
+	dimensions: {
+		type: 'object',
+	},
```
Why do we need to store the dimensions? Can't we just use percentage in the focal point instead?

jeffersonrabb replied:
The focal point picker needs image dimensions for two reasons:

- To define the drag area for the target. The user should only be able to drag the target over the area of the image, and unless the image's aspect ratio precisely matches the focal point picker's area there will be blank space on the left/right or top/bottom. Awareness of the dimensions allows a precise definition of the drag area. This is done here: https://github.com/WordPress/gutenberg/blob/97fcd04d05ce2ae039f891225a6a1fb781edfab2/packages/components/src/focal-point-picker/index.js#L39-L64
- To convert the location of the target to a percentage-based focal point we need to know the dimensions of the original image. This calculation occurs here: https://github.com/WordPress/gutenberg/blob/97fcd04d05ce2ae039f891225a6a1fb781edfab2/packages/components/src/focal-point-picker/index.js#L65-L90

However, I don't love requiring dimensions as a parameter for the component. First of all, it's restrictive - ideally the component just takes the image path and handles the rest. And second, it also creates a deprecation problem for Cover blocks created prior to the Focal Point Picker in which the dimensions are not retained. Ideally, I'd prefer that the Focal Point Picker component determine the dimensions programmatically from the image itself. I'm going to experiment with a few approaches to this and will update the PR shortly.

jeffersonrabb replied:
In https://github.com/WordPress/gutenberg/pull/10925/commits/5567d1649dfde7a289362df23aa09e2c884ed109 I've introduced a different approach to getting the image dimensions. The parameters are gone. Instead, the component creates an invisible `<img />` (opacity: 0) and derives the dimensions from that. This way the block using the component need not supply dimensions and the component will work with legacy blocks. Does this seem like a viable approach?

jasmussen commented:
Really nice work.

GIF:

![focalpoint](https://user-images.githubusercontent.com/1204802/51376191-05a68c00-1b08-11e9-89fe-f91ebd936a0d.gif)

Can you make the focal point picker look like this?

![group 2](https://user-images.githubusercontent.com/1204802/51376236-2242c400-1b08-11e9-823e-617e17aa4d8e.png)

That is, `$blue-medium-600` for the spot color, 1px white stroke around it so it's visible on all backgrounds, and 2px thick circle. This makes it feel related to the resize handles on images:

<img width="478" alt="screenshot 2019-01-18 at 10 03 55" src="https://user-images.githubusercontent.com/1204802/51376348-6635c900-1b08-11e9-9268-bb5220ee1eda.png">

There's white padding above and below the focal point picker — is this intended? Can it be removed?

jasmussen commented:
Can you add input fields below the picker that matches the percentages typed in? This would be an accessible option:

![group 2](https://user-images.githubusercontent.com/1204802/51376638-173c6380-1b09-11e9-93c9-713ff1b0cd5b.png)

jeffersonrabb committed:
Display horizontal and vertical positions as percentages in text boxes, for accessibility reasons. Slight corrections to the postion calculations.

jeffersonrabb commented:
> Can you add input fields below the picker that matches the percentages typed in? This would be an accessible option:

Done in https://github.com/WordPress/gutenberg/pull/10925/commits/7131d07f8752938adbd6aee0c5ca3d375102dc12. I added my own styles for placing the two TextControls side-by-side. If there is a Gutenberg component or other structure that makes this easier to do, let me know and I will replace. Also note that these text fields are not editable. Making them editable would introduce a bit more complexity - is this needed?

jeffersonrabb committed:
Alternate approach to getting image dimensions, without requiring that they be passed in as props. Create an invisible image element and get dimensions from it.

jeffersonrabb committed:
Removing dimensions attribute from Cover block.

kjellr commented:
This is fantastic. Looking forward to having it in. 

> Making them editable would introduce a bit more complexity - is this needed?

Yeah, I believe so. I just tried changing the focal point with only the keyboard, and was unable to. Editing those fields would add that functionality. 

Also, before the focal point had been adjusted, I noticed the field were showing up as `NaN%` for me: 

<img width="279" alt="screen shot 2019-01-18 at 1 15 49 pm" src="https://user-images.githubusercontent.com/1202812/51405178-9fc7fd80-1b23-11e9-93e4-9218af82a639.png">

Those could probably either both be `50%`, or blank.

jeffersonrabb committed:
Fix to bug that displayed NaN/NaN in position fields before target is moved.

jeffersonrabb commented:
> Yeah, I believe so. I just tried changing the focal point with only the keyboard, and was unable to. Editing those fields would add that functionality.

If the TextControls are going to be editable, for one thing we'll need to scrap the percentage sign, or move it outside of the text fields. Does this seem problematic for any reason?

jeffersonrabb committed:
Make text input fields for horizontal and vertical position editable.

jeffersonrabb commented:
@kjellr I've made the text fields editable in https://github.com/WordPress/gutenberg/pull/10925/commits/8bd268d633958c6920d98176c335ccfb329ab9e8. To do this I had to pull the percentage signs out of the fields. I attempted to render them outside the input using an `::after` pseudo class, but I'm not having any luck. Will work on this a bit more shortly.

jeffersonrabb committed:
Completely replace background image with IMG tag approach.

jeffersonrabb committed:
Calculate bounds only when image onLoad fires.

jeffersonrabb committed:
Replace TextControl elements with plain inputs. Percentage sign after text fields.

jeffersonrabb commented:
> I attempted to render them outside the input using an `::after` pseudo class, but I'm not having any luck.

Percentage signs in as of https://github.com/WordPress/gutenberg/pull/10925/commits/48d494242f8b4257cfb8d6b82ffe92812ba078dc

jeffersonrabb committed:
Adding a bit of padding around the image, so the white space that occurs based on aspect ratio variance doesn't track as an error.

jeffersonrabb commented:
> There's white padding above and below the focal point picker — is this intended? Can it be removed?

This is intentional—by keeping the size of the component fixed and displaying the complete image within it we protect against an uncomfortably tall or short component size if the image is either narrow/tall (e.g. image of a skyscraper) or wide/short (e.g. a panorama). Per a suggestion from @jasmussen I've added some padding around the image so the whitespace doesn't track as an error. This is done in https://github.com/WordPress/gutenberg/pull/10925/commits/87b63dbc513fa252f6ce1349f86dc97842e8a1a7

Commit:
Small a11y and visual fixes

- Change tab index of focus picker to -1 so it's not tabbable, it's not actionable by keyboard anyway.
- Change padding on input fields to better fit the percentages.
- Small syntax cleanups.

Commit:
Use an SVG, and theme the colors

- Change the ring to be an SVG. This allows us to change the colors to match the theme.
- Add draggable false property to thumbnail, so you don't accidentally copy it to your desktop if you start the dragging operation outside the ring handle
- Redesign the drag handle, by the way. Ring seems sufficient.
- Increase the opacity a bit.
- Change some of the CSS to leverage PostCSS rewriting.

jasmussen commented:
Pushed a few fixes.

- Change tab index of focus picker to -1 so it's not tabbable, it's not actionable by keyboard anyway.
- Change padding on input fields to better fit the percentages.
- Small syntax cleanups.
- Change the ring to be an SVG. This allows us to change the colors to match the theme.
- Add draggable false property to thumbnail, so you don't accidentally copy it to your desktop if you start the dragging operation outside the ring handle
- Redesign the drag handle, by the way. Ring seems sufficient.
- Increase the opacity a bit.
- Change some of the CSS to leverage PostCSS rewriting.

GIFs:

![default colors](https://user-images.githubusercontent.com/1204802/51466776-47833c80-1d6b-11e9-8e82-90e379367565.gif)

And if you change to the Midnight theme:

![themed colors](https://user-images.githubusercontent.com/1204802/51466793-4eaa4a80-1d6b-11e9-9612-b7ed97c461e5.gif)

jasmussen commented:
With the recent changes, this is good to go from a design & accessibility perspective. Both could use sanity checks, of course, but thumbs up from me.

youknowriad reviewed (commented):
Nice work here, left some minor notes.

kjellr commented:
> With the recent changes, this is good to go from a design & accessibility perspective.

+1 to this. The only small issue I'm seeing now is there are a lot of selection artifacts left over when I drag the focal point picker around in Safari: 

![ghost](https://user-images.githubusercontent.com/1202812/51482128-95944200-1d63-11e9-8bea-533b632fe5be.gif)

Not sure if there's anything we can do about that.

jeffersonrabb commented:
> The only small issue I'm seeing now is there are a lot of selection artifacts left over when I drag the focal point picker around in Safari:

Before I dig in on this, do you happen to recall if the issue was occurring last week, before @jasmussen replaced the target SVG?

jasmussen commented:
Oh hey, one thing springs to mind here. There were some `translate` styles baked into the SVG background I replaced, those almost certainly have something to do with it:

https://github.com/WordPress/gutenberg/pull/10925/commits/1bd8ea03c4eb65b52fdb5dd1fbb578842e2d7912#diff-1315a70baa9fba9ba6182808f2bbe75bL1

`translate(1 1)`

Not sure what that does, specifically, but I'm familiar with an old hack where we apply `transform: translateZ(0)` on an object we want to be painted by the GPU for performance. This is likely what that old rule did as well, so perhaps worth trying. @kjellr does that ring any bells?

kjellr commented:
> Oh hey, one thing springs to mind here. There were some translate styles baked into the SVG background I replaced, those almost certainly have something to do with it:
> 
> 1bd8ea0#diff-1315a70baa9fba9ba6182808f2bbe75bL1
> 
> translate(1 1)
> 
> Not sure what that does, specifically, but I'm familiar with an old hack where we apply transform: translateZ(0) on an object we want to be painted by the GPU for performance. This is likely what that old rule did as well, so perhaps worth trying. @kjellr does that ring any bells?

Yep, that's the issue. Adding `translateZ(0)` or `will-change: transform;` eliminates the issue.

jeffersonrabb committed:
CSS tweak to avoid visual artifacts in Safari during dragging.

jeffersonrabb committed:
Changed "component" prefix for all classes to "components" to match convention. Moved render() function to be the last one.

jeffersonrabb commented on packages/components/src/focal-point-picker/index.js:
```diff
@@ -0,0 +1,251 @@
+			top: `${ iconCoordinates.top }px`,
+		};
+		const iconContainerClasses = classnames(
+			'component-focal-point-picker__icon_container',
```
Fixed in https://github.com/WordPress/gutenberg/pull/10925/commits/d61d15afa275c9aa9c0663bb9d431773b6cff184

jeffersonrabb commented on packages/components/src/focal-point-picker/index.js:
```diff
@@ -0,0 +1,251 @@
+			</BaseControl>
+		);
+	}
+	handleFocusOutside() {
```
Done in https://github.com/WordPress/gutenberg/pull/10925/commits/d61d15afa275c9aa9c0663bb9d431773b6cff184

jeffersonrabb commented:
> Yep, that's it. Adding `translateZ(0)` or `will-change: transform;` eliminates the issue.

Added `will-change:transform;` in https://github.com/WordPress/gutenberg/pull/10925/commits/b44f1b9de2d5a62632fd099d452141d032e5a676

Seems to work well.

kjellr commented:
> Added will-change:transform; in b44f1b9
> 
> Seems to work well.

Can confirm — I'm no longer seeing those artifacts. 

I think we're all set from a design perspective now. 👍

jasmussen commented:
🎉

afercia commented:
I've just noticed this component renders two input fields that are completely unlabelled. The visible text is just plain text.

There's no associated `<label>` element or `aria-label` attribute to give these two controls a proper label. I'm going to create an issue.

Similar cases happened several times in the past months and I'm sorry to have to point it out again: this is a very basic accessibility requirement. There's been some discussion in the past about a way to try to avoid this kind of oversights but nothing concrete happened. Please, let's try to find together a way to improve the process. 🙂

aduth commented on packages/components/src/focal-point-picker/index.js:
```diff
@@ -0,0 +1,251 @@
+					</div>
+				</div>
+				<div className="components-focal-point-picker_position-display-container">
+					<BaseControl label={ __( 'Horizontal Pos.' ) }>
```
Per @afercia 's comment, I think this needed to have been assigned an `id` for a label to be produced associated with the rendered `input` (which _does_ have an `id` associated).

It seems at some point `BaseControl` was updated to be accommodating to receiving a `label` and not an `id`, at which point it renders a `span`. I don't know that this should be supported, as it leaves open the possibility for errors like this.

https://github.com/WordPress/gutenberg/blob/858719bf3e5c7cd37c14824f5f3d4b8db49cc8f2/packages/components/src/base-control/index.js#L11

afercia replied:
Thanks @aduth I think that change was made for the new font size picker:

<img width="447" alt="screenshot 2019-02-15 at 15 23 32" src="https://user-images.githubusercontent.com/1682452/52862544-cb93ce80-3135-11e9-8b83-6e32cbc08ef8.png">

where the visible text can't be a label element. At that point though, the input child _must_ have an `aria-label` attribute: and yes, this leaves open the possibility that developers forget or don't know they need to label the input.

aduth replied:
> where the visible text can't be a label element

(Related: https://github.com/WordPress/gutenberg/pull/9802#discussion_r257250785 )

Forgive my understanding, but it's not clear to me why in that example "Font Size" would not be a label element associated with the dropdown input?

afercia replied:
Not all elements can have an associated <label> element. Form elements and a few others can. What is the dropdown input in this case? The div?

aduth replied:
> Not all elements can have an associated element. Form elements and a few others can. What is the dropdown input in this case? The div?

There's a `button` which could receive the initial focus for the dropdown.

![image](https://user-images.githubusercontent.com/1779930/52866415-9921ab80-310c-11e9-9344-a1905b8f7239.png)

afercia replied:
Yes, technically a button is a labelable element, here's a list: https://www.w3.org/TR/html52/sec-forms.html#labelable-element

However, it's not just about click > focus. The primary goal of labels is giving names to controls. In this case, it wouldn't make sense. A `<label>` element would override the button text. It wouldn't override the button aria-label though, which is, for example: `aria-label: "Font size: Normal"`. It would be a very uncommon usage. Seems to me your proposal in https://github.com/WordPress/gutenberg/pull/9802#discussion_r257269125 is preferable.

aduth replied:
FYI these specific `BaseControl` were corrected by @jorgefilipecosta in #14152

swissspidy commented:
Why wasn't the addition of this new component reflected in the changelog?

aduth commented:
> Why wasn't the addition of this new component reflected in the changelog?

It should have been, but the changelog workflows are not universally adopted / enforced. This should become less an issue after #13594.
